usb: cdc_acm: allow to configure CDC ACM UART device from devicetree

jfischer-no opened the pull request:
Add hidden Kconfig option to Kconfig.cdc and allow
to configure CDC ACM UART device from devicetree.

jfischer-no committed:
usb: cdc_acm: allow to configure CDC ACM UART device from devicetree

Add hidden Kconfig option to Kconfig.cdc and allow
to configure CDC ACM UART device from devicetree.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: usb: console: get CDC ACM UART device from devicetree

Add app.overlay which contains chosen node and cdc-acm-uart node.
Rework sample to get CDC ACM UART device from devicetree.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: usb: console: convert README to RST file

Convert README to RST file.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

mbolivar-nordic commented on subsys/usb/class/cdc_acm.c:
```diff
@@ -1118,4 +1118,19 @@ static const struct uart_driver_api cdc_acm_driver_api = {
+DT_INST_FOREACH_STATUS_OKAY(CDC_ACM_DT_DEVICE_DEFINE);
+#else
 UTIL_LISTIFY(CONFIG_USB_CDC_ACM_DEVICE_COUNT, CDC_ACM_DEVICE_DEFINE, _)
+#endif
```
Any reason not to allow both ways of defining these devices?

jfischer-no replied:
I prefer only one option, as it is with other UART drivers for hardware controllers, and I would like to deprecate Kconfig/UTIL_LISTIFY option. I also see no advantage but a lot of conflict potential in the configuration for the user.

carlescufi replied:
I agree. It would be quite confusing if the user can specify both `CONFIG_USB_CDC_ACM_DEVICE_COUNT=2` and then add a 3rd USB CDC port via Devicetree. I think that keeping the old Kconfig config for backwards compatibility but switching to Devicetree entirely whenever it is used to define these ports makes more sense.

carlescufi replied:
In fact, ideally this PR would be extended to convert the in-tree samples that use Kconfig to DT, and deprecate the Kconfig options themselves. We shouldn't have two ways of achieving exactly the same objective, and DT is more flexible in that it allows the user to change settings per-instance, as well as being able to connect the `chosen` node for console, shell, logging, etc directly to a real DT node instead of falling back to Kconfig.

mbolivar-nordic replied:
I agree. Let's deprecate the Kconfig way of doing things.

@jfischer-no is USB CDC ACM the only reason we have CONFIG_UART_CONSOLE_ON_DEV_NAME?

If so, that should be deprecated too. That would help avoid the weird issues in the last patch, since we could:

- just get the console device out of devicetree directly with `DEVICE_DT_GET(DT_CHOSEN(zephyr_console))` to poll DTR
- verify the console is USB by using `BUILD_ASSERT(DT_NODE_HAS_COMPAT(DT_CHOSEN(zephyr_console)), zephyr_cdc_acm_uart), ...);`

While we are doing this, I wonder if we can deprecate the following Kconfig options in the same way if CDC ACM is the only thing keeping them around:

- CONFIG_BT_MONITOR_ON_DEV_NAME
- CONFIG_BT_UART_ON_DEV_NAME
- CONFIG_UART_SHELL_ON_DEV_NAME
- CONFIG_UART_PIPE_ON_DEV_NAME

jfischer-no replied:
> @jfischer-no is USB CDC ACM the only reason we have CONFIG_UART_CONSOLE_ON_DEV_NAME?

There is also: `boards/nios2/altera_max10/doc/index.rst:   CONFIG_UART_CONSOLE_ON_DEV_NAME="jtag_uart0"`, `drivers/serial/uart_altera_jtag_hal.c`

jfischer-no replied:
[subsys/bluetooth/host/monitor.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/monitor.c#L237) needs to be changed as well.

mbolivar-nordic commented on samples/subsys/usb/console/app.overlay:
```diff
@@ -0,0 +1,17 @@
+	};
+
+	cdc_acm_uart0: uart {
+		status = "okay";
```
You can leave out the 'status = "okay";' if you want, since that is the default value. But it's your decision.

mbolivar-nordic commented on samples/subsys/usb/console/src/main.c:
```diff
@@ -6,29 +6,30 @@
+	/* Poll if the DTR flag was set */
 	while (!dtr) {
 		uart_line_ctrl_get(dev, UART_LINE_CTRL_DTR, &dtr);
+		/* Give CPU resources to low priority threads. */
```
Confusing comment here. Is there another thread in the system? I don't see the application starting any; does the USB subsystem start some threads automatically?

pdunaj replied:
it does, USB has its own workqueue

mbolivar-nordic replied:
Thanks, resolving

mbolivar-nordic commented on samples/subsys/usb/console/src/main.c:
```diff
@@ -6,29 +6,30 @@
-	if (strlen(CONFIG_UART_CONSOLE_ON_DEV_NAME) !=
-	    strlen("CDC_ACM_0") ||
-	    strncmp(CONFIG_UART_CONSOLE_ON_DEV_NAME, "CDC_ACM_0",
+	if (strlen(CONFIG_UART_CONSOLE_ON_DEV_NAME) != strlen(CDC_ACM_LABEL) ||
+	    strncmp(CONFIG_UART_CONSOLE_ON_DEV_NAME, CDC_ACM_LABEL,
```
I'm confused about what's going on here. If you're polling DTR on `dev`, then I think the right comparison is that `dev->name` is `CONFIG_UART_CONSOLE_ON_DEV_NAME` instead. Although that still isn't 100% correct since there could be multiple CDC ACM devices, and `dev` is one, but it isn't the console.

pdunaj replied:
there is no point in using `strncmp` as `strlen` that is used for limit is called with the same argument

pdunaj replied:
but both issues seems historical

jfischer-no replied:
AFAIK the idea was to catch the error if chosen node and CONFIG_UART_CONSOLE_ON_DEV_NAME has the default value from board's dts. But this is only the case if someone changes this sample to try something, e.g. remove or rename chosen node in samples app.overlay.

I can remove it if desired.

mbolivar-nordic replied:
> I can remove it if desired.

See above for a suggested solution.

carlescufi reviewed (approved):
Looks great to me!

emob-nordic commented:
Please DO NOT merge. Im currently reviewing.

eobalski reviewed (changes requested):
Nacking to avoid merge, Im reviewing.

emob-nordic commented on subsys/usb/class/Kconfig.cdc:
```diff
@@ -1,15 +1,26 @@
+
+menu "USB CDC ACM Class support"
+
+config USB_CDC_ACM_HAS_DT_COMPAT
+	bool
+	default $(dt_compat_enabled,$(DT_COMPAT_ZEPHYR_CDC_ACM_UART))
+	imply USB_CDC_ACM
+	help
+	  Helper symbol to select USB_CDC_ACM if compatible node is enabled.
```
I thought its better to have each class as menuconfig.
What is the advantage of having this hidden option vs having:

``` patch
diff --git a/subsys/usb/class/Kconfig.cdc b/subsys/usb/class/Kconfig.cdc
index bc9e169afa..f7e9d96465 100644
--- a/subsys/usb/class/Kconfig.cdc
+++ b/subsys/usb/class/Kconfig.cdc
@@ -3,21 +3,13 @@
 
 DT_COMPAT_ZEPHYR_CDC_ACM_UART := zephyr,cdc-acm-uart
 
-menu "USB CDC ACM Class support"
-
-config USB_CDC_ACM_HAS_DT_COMPAT
-	bool
-	default $(dt_compat_enabled,$(DT_COMPAT_ZEPHYR_CDC_ACM_UART))
-	imply USB_CDC_ACM
-	help
-	  Helper symbol to select USB_CDC_ACM if compatible node is enabled.
-
-config USB_CDC_ACM
+menuconfig USB_CDC_ACM
 	bool "Enable USB CDC ACM Class support"
 	select SERIAL_HAS_DRIVER
 	select SERIAL_SUPPORT_INTERRUPT
 	select RING_BUFFER
 	select UART_INTERRUPT_DRIVEN
+	default $(dt_compat_enabled,$(DT_COMPAT_ZEPHYR_CDC_ACM_UART))
 	help
 	  Enable USB CDC ACM Class support.
 	  Default UART device name is "CDC_ACM_0".
@@ -37,7 +29,7 @@ config USB_CDC_ACM_DEVICE_NAME
 	  Device name template for the CDC ACM Devices. First device would
 	  have name $(USB_CDC_ACM_DEVICE_NAME)_0, etc.
 
-if !USB_CDC_ACM_HAS_DT_COMPAT
+if !$(dt_compat_enabled,$(DT_COMPAT_ZEPHYR_CDC_ACM_UART))
 module = USB_CDC_ACM
 default-count = 1
 source "subsys/usb/class/Kconfig.template.composite_device_number"
@@ -77,5 +69,3 @@ module-str = usb cdc acm
 source "subsys/logging/Kconfig.template.log_config"
 
 endif # USB_CDC_ACM
-
-endmenu
```

The only thing that's different is not the best description for why is `USB_CDC_ACM` y selected, but is that really an issue?

jfischer-no replied:
> I thought its better to have each class as menuconfig.
> What is the advantage of having this hidden option vs having:

It still has a menu because of `menu "USB CDC ACM Class support"`.
It is for debugging, USB_CDC_ACM_HAS_DT_COMPAT allows to see what _y-selecting this symbol_ or _y-implying_, here USB_CDC_ACM.

emob-nordic replied:
There are many examples already in the tree that does the same without helper option. I think the solution i proposed is much more compact and elegant. I believe zephyr users are mature enough to understand why this is `y` implied by looking into the Kconfig directly. To clarify it you could add some note in help
```
 	help
 	  Enable USB CDC ACM Class support.
 	  Default UART device name is "CDC_ACM_0".
 	  This option is automatically enabled if USB CDC ACM
 	  is chosen as zephyr console in DTS.
```

Menuconfig option could be selected, menu requires to go one step deeper into configuration gui. I wouldn't change that.

jfischer-no replied:
I disagree.

jfischer-no replied:
USB_CDC_ACM_HAS_DT_COMPAT is removed in the commit "usb: cdc_acm: configure CDC ACM UART instances using devicetree only"

emob-nordic commented on subsys/usb/class/Kconfig.cdc:
```diff
@@ -1,15 +1,26 @@
 	help
-	  USB CDC ACM device class support.
-	  Default device name is "CDC_ACM_0".
+	  Enable USB CDC ACM Class support.
```
Every class has something like `USB <class_name> (Device) Class support` except for CDC ACM right now.

emob-nordic replied:
I ment: remove `"Enable"`

eobalski reviewed (changes requested):
Im a fan of the change.
Most of my comments are suggestions i see valuable if you want to clean up things and make them compact.
Would the next step be to enforce users to use dt over Kconfig? This allows to have separate configurations for few CDCs in composite configuration. Plus removes `UTIL_LISTIFY` :)

mbolivar-nordic reviewed (changes requested):
-1 since I think CONFIG_UART_CONSOLE_ON_DEV_NAME and potentially additional Kconfigs should be deprecated, but I am in favor of where this is going.

jfischer-no committed:
samples: shell: get CDC ACM UART device from devicetree

Add usb.overlay which contains chosen node and cdc-acm-uart node.
Change sample to get CDC ACM UART device from devicetree.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
net: ot: rework NCP interface configuration

Rework NCP interface configuration and NCP sample. Remove
CONFIG_OPENTHREAD_COPROCESSOR_SPINEL_ON_UART_DEV_NAME and
CONFIG_OPENTHREAD_COPROCESSOR_SPINEL_ON_UART_ACM Kconfig
options in favor of chosen node zephyr,ot-uart usage.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: wpan_serial: get CDC ACM UART device from devicetree

Add app.overlay which contains  cdc-acm-uart node.
Rework sample to get CDC ACM UART device from devicetree.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: usb: get CDC ACM UART device from devicetree

Add app.overlay which contains cdc-acm-uart nodes.
Rework CDC ACM and HID samples to get CDC ACM UART device
from devicetree.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: sensortile_box: get chosen console device from devicetree

Add app.overlay which contains chosen node
Rework sample to get CDC ACM UART device from devicetree.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
soc: arc: remove unnecessary UART_CONSOLE_ON_DEV_NAME option

This configuration is also available on board level by
chosen node "zephyr,console".

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
boards: degu_evk: add chosen nodes for shell and console

Remove UART_CONSOLE_ON_DEV_NAME and UART_SHELL_ON_DEV_NAME options
and add chosen nodes for shell and console.

Enable USB device support because the board uses
CDC ACM UART for console.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
tests: uart: remove CONFIG_UART_CONSOLE_ON_DEV_NAME usage

Use DT_CHOSEN(zephyr_console) instead of
CONFIG_UART_CONSOLE_ON_DEV_NAME Kconfig option.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
console: remove device_get_binding(CONFIG_UART_CONSOLE_ON_DEV_NAME)

Remove device_get_binding(CONFIG_UART_CONSOLE_ON_DEV_NAME)
and use DEVICE_DT_GET(DT_CHOSEN(zephyr_console)) to get
chosen "zephyr,console" node.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: remove Kconfig option CONFIG_UART_CONSOLE_ON_DEV_NAME

Remove Kconfig option CONFIG_UART_CONSOLE_ON_DEV_NAME.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

galak commented on doc/reference/devicetree/api.rst:
```diff
@@ -319,7 +319,7 @@ device.
    * - zephyr,console
-     - Sets default :kconfig:`CONFIG_UART_CONSOLE_ON_DEV_NAME`
+     - Sets UART device used by console driver
    * - zephyr,dtcm
```
Add all the new zephyr,ot-uart, etc. here.

mbolivar-nordic replied:
> Add all the new zephyr,ot-uart, etc. here.

Still needs to be done

jfischer-no replied:
See commit "net: ot: rework NCP interface configuration", https://github.com/zephyrproject-rtos/zephyr/pull/37085/commits/7d4ec15500d7b2b34fd31c03cd16d8f8e8e1f946#diff-4d66c0cb52a8c2ce6a8cf7e5f0b73fbdca77ed0f93e9f9b0d810471cd49e3ddaR342-R343

de-nordic commented:
We need to check how it will impact mcuboot serial recovery, which currently uses the `zephyr,console` as a default port.
I have already tried to change that (https://github.com/mcu-tools/mcuboot/pull/951); with the addition of this PR we should probably also remove the Kconfig options there and relay on the DTS.

jfischer-no committed:
tests: uart_basic_api: add support for CDC ACM UART

Add support for CDC ACM UART.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
tests: uart_basic_api: allow USB workqueue/thread to work

Allow UART driver that offload interrupt processing
to workuqueue, like CDC ACM UART to work.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
boards: rak5010_nrf52840: remove wrong uart1 node compatible

uart1 node compatible should not be changed.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
boards: nrf52840dongle_nrf52840: do not default SERIAL on USB_CDC_ACM

Prevent configuration dependency loop.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
bluetooth: monitor: remove CONFIG_UART_CONSOLE_ON_DEV_NAME

Replace with DT_LABEL(DT_CHOSEN(zephyr_console)).

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: cdc_acm: configure CDC ACM UART instances using devicetree only

Remove the possibility to configure number of CDC ACM UART
instances in Kconfig.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
manifest: MCUboot update to get CDC ACM device from devicetree

MCUboot version update to get CDC ACM device from devicetree.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
tests: bluetooth: shell: update shell over CDC ACM UART configuration

Add usb.overlay which contains chosen node and cdc-acm-uart node.
Update USB configuration and test case.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: uart_mcumgr: remove CONFIG_UART_MCUMGR_ON_DEV_NAME

Remove CONFIG_UART_MCUMGR_ON_DEV_NAME and use
DEVICE_DT_GET(DT_CHOSEN(zephyr_uart_mcumgr)).
Add usb.overlay, which contains chosen node and cdc-acm-uart node,
to smp_svr sample.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

carlescufi commented:
@emob-nordic could you please take another look?

mbolivar-nordic commented on samples/boards/sensortile_box/app.overlay:
```diff
@@ -0,0 +1,17 @@
+		zephyr,console = &cdc_acm_uart1;
+	};
+
+	cdc_acm_uart1: cdc_acm_uart@1 {
+		compatible = "zephyr,cdc-acm-uart";
+		reg = <0x1 0x1>;
+		label = "CDC_ACM_1";
+	};
```
@jfischer-no Let's bring the discussion we started offline about this into the PR.

Problem statement: this way of representing the CDC ACM configuration in devicetree does not create a link between the zephyr,cdc-acm-uart node and the node which represents the USB device controller hardware. This creates ambiguity if multiple USB device controllers are present and enabled in the same system. Specifically the ambiguity is: which USB device should be used for CDC ACM communication?

You told me offline that there is a plan to add support for multiple USB controllers to Zephyr's USB subsystem. Therefore, I think we should solve this problem to avoid forcing users to change their devicetrees later.

I can see two options:

1. Make these zephyr,cdc-acm-uart nodes children of the USB node. Example for Nordic:

```
&usbd {
	cdc_acm_uart1: cdc_acm_uart@1 {
		compatible = "zephyr,cdc-acm-uart";
		reg = <0x1 0x1>;
		label = "CDC_ACM_1";
	};
};
```

Then I think we should add `bus: usb` to the bindings for each USB controller. E.g. add `bus: usb` to nordic,nrf-usbd.yaml.

Then from the CDC ACM driver you can say:

```
#define DT_DRV_COMPAT zephyr_cdc_acm_uart

...
DT_BUS(DT_DRV_INST(inst))
```

if you need a node identifier for the USB controller.

2. With a phandle. E.g. something like this:

```
	cdc_acm_uart1: cdc_acm_uart@1 {
		compatible = "zephyr,cdc-acm-uart";
		reg = <0x1 0x1>;
		label = "CDC_ACM_1";
		usb-device = <&usbd>;
	};
```

And adding a `usb-device` property to the binding, with `type: phandle`.

And then use `DT_PHANDLE(DT_DRV_INST(inst), usb_device)` to get the node identifier for the USB controller.


What do you think? Any other ideas? @galak?

emob-nordic replied:
Property name `usb-device` is unfortunate in USB nomenclature. I would opt for `controller`.

jfischer-no replied:
I follow the approach "Make these zephyr,cdc-acm-uart nodes children of the USB node"

mbolivar-nordic commented on samples/subsys/usb/hid-cdc/src/main.c:
```diff
@@ -567,13 +567,13 @@ void main(void)
 
-	cdc0_dev = device_get_binding("CDC_ACM_0");
+	cdc0_dev = DEVICE_DT_GET(DT_INST(0, zephyr_cdc_acm_uart));
 	if (cdc0_dev == NULL) {
 		LOG_ERR("Cannot get USB CDC 0 Device");
 		return;
 	}
 
-	cdc1_dev = device_get_binding("CDC_ACM_1");
+	cdc1_dev = DEVICE_DT_GET(DT_INST(1, zephyr_cdc_acm_uart));
 	if (cdc1_dev == NULL) {
```
The same comment that applies to the cdc_acm_composite sample applies here.

mbolivar-nordic commented on samples/net/wpan_serial/app.overlay:
```diff
@@ -0,0 +1,14 @@
+ */
+
+/ {
+	usbd0 {
```
Does this really work? I thought aliases could not be used this way in your testing. The same question applies to other app.overlay files in this PR that do the same thing.

jfischer-no replied:
That works, what does not work is to define cdc_acm_uart0 node as a child of a node using reference and enable it here.

mbolivar-nordic commented on samples/subsys/usb/console/src/main.c:
```diff
@@ -6,33 +6,26 @@
 {
-	const struct device *dev = device_get_binding(
-			CONFIG_UART_CONSOLE_ON_DEV_NAME);
+	const struct device *dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_console));
```
It is excellent that this sample can be reworked in this way with the latest update.

mbolivar-nordic commented on tests/drivers/uart/uart_basic_api/src/test_uart_config.c:
```diff
@@ -36,7 +36,7 @@ const struct uart_config uart_cfg = {
 static int test_configure(void)
 {
-	const struct device *uart_dev = device_get_binding(UART_DEVICE_NAME);
+	const struct device *uart_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_console));
 
 	if (!uart_dev) {
```
The `!uart_dev` change to `!device_is_ready(uart_dev)` needs to be done for all the changes in this commit as well.

mbolivar-nordic commented on drivers/console/uart_console.c:
```diff
@@ -597,7 +597,7 @@ static int uart_console_init(const struct device *arg)
 
 	/* Claim console device */
-	uart_console_dev = device_get_binding(CONFIG_UART_CONSOLE_ON_DEV_NAME);
+	uart_console_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_console));
```
Should this driver return -ENODEV if this device is not ready?

mbolivar-nordic commented on subsys/console/getchar.c:
```diff
@@ -52,7 +52,7 @@ int console_init(void)
 	int ret;
 
-	uart_dev = device_get_binding(CONFIG_UART_CONSOLE_ON_DEV_NAME);
+	uart_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_console));
 	ret = tty_init(&console_serial, uart_dev);
```
Any device_is_ready changes needed here?

mbolivar-nordic commented on subsys/logging/log_backend_uart.c:
```diff
@@ -75,7 +75,7 @@ static void process(const struct log_backend *const backend,
 {
-	uart_dev = device_get_binding(CONFIG_UART_CONSOLE_ON_DEV_NAME);
+	uart_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_console));
 	__ASSERT_NO_MSG((void *)uart_dev);
```
`__ASSERT_NO_MSG(device_is_ready(uart_dev))`

mbolivar-nordic commented on tests/drivers/uart/uart_basic_api/src/main.c:
```diff
@@ -57,6 +58,20 @@ void test_uart_pending(void)
+	const struct device *dev;
+	uint32_t dtr = 0;
+
+	dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_console));
+	if (dev == NULL || usb_enable(NULL)) {
```
device_is_ready

mbolivar-nordic commented on boards/arm/rak5010_nrf52840/rak5010_nrf52840.dts:
```diff
@@ -75,7 +75,7 @@
 };
 
 &uart1 {
-	compatible = "nordic,nrf-uart";
+	compatible = "nordic,nrf-uarte";
```
It would be better just to remove this line completely. The peripheral does not have the option to be reconfigured as UART, so just let nrf52840.dtsi provide the correct value for compatible.

mbolivar-nordic commented on subsys/usb/class/cdc_acm.c:
```diff
@@ -1074,19 +1074,6 @@ static const struct uart_driver_api cdc_acm_driver_api = {
 		.tx_ringbuf = &tx_ringbuf_##x,				\
 	};
 
-#define CDC_ACM_DEVICE_DEFINE(idx, _)					\
```
Hurray!

mbolivar-nordic commented on drivers/console/uart_mcumgr.c:
```diff
@@ -182,7 +182,7 @@ void uart_mcumgr_register(uart_mcumgr_recv_fn *cb)
 	uart_mgumgr_recv_cb = cb;
 
-	uart_mcumgr_dev = device_get_binding(CONFIG_UART_MCUMGR_ON_DEV_NAME);
+	uart_mcumgr_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_uart_mcumgr));
 
 	if (uart_mcumgr_dev != NULL) {
```
device_is_ready

mbolivar-nordic reviewed (changes requested):
Thanks! Some comments about this latest update.

jfischer-no commented on dts/bindings/serial/zephyr,cdc-acm-uart.yaml:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2021 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+description: USB CDC ACM UART
```
TBH, I do not want to do it now because there is no need for it. We do not have support for multiple controllers. And when the time comes, the CDC ACM classe will have to be re-implemented.

mbolivar-nordic replied:
> TBH, I do not want to do it now because there is no need for it.

I think if you added this, plus the BUILD_ASSERT I described [below](https://github.com/zephyrproject-rtos/zephyr/pull/37085#discussion_r683861679), you would have got build errors for the invalid use of aliases. So I disagree there is no need for it -- it can be used to catch programmer errors in the DTS files.

jfischer-no commented on boards/arm/96b_aerocore2/96b_aerocore2.dts:
```diff
@@ -36,6 +36,7 @@
 	aliases {
 		led0 = &yellow_led_1;
 		led1 = &blue_led_2;
+		usbd0 = &usbotg_fs;
```
DT_ALIAS(usbd0) is not a reason, usbd0 is common aliases for the USB device controller node and is used in all the samples overlay files.

jfischer-no replied:
This aliases actually has no effect, I think it is not possible to use them like this. app.overlay below works just as well.
```
/ {                                                                                           
        idonotlikealiases0 {                                                                               
                cdc_acm_uart0 {                                                               
                        compatible = "zephyr,cdc-acm-uart";                                   
                        label = "CDC_ACM_0";                                                  
                };                                                                            
        };                                                                                    
}; 
```

mbolivar-nordic replied:
> This aliases actually has no effect, I think it is not possible to use them like this. app.overlay below works just as well.

Right, so this is creating a node `/idonotlikealiases0/cdc_acm_uart0`, which is not a child of any USB controller. That is the problem that you can avoid by implementing the `bus: usb` + `on-bus: usb` suggestion I posted about above:

https://github.com/zephyrproject-rtos/zephyr/pull/37085#discussion_r682974127

Unfortunately, due to the way the DT tooling works right now, you cannot use `on-bus: usb` to force the node to be on that bus, because that is also used to make multiple bindings for the same compatible on different buses (like separate bindings for SPI and I2C for the same compatible).

However, you can do something like this:

```
#define CDC_ACM_DT_DEVICE_DEFINE(idx)					\
	BUILD_ASSERT(DT_INST_ON_BUS(idx, usb),				\
		     "node " DT_NODE_PATH(DT_DRV_INST(idx))		\
		     " is not on a USB bus");				\
	...
```

Than you would know that the nodes are all on the right sort of bus, or get a sensible error, at build time.

jfischer-no replied:
that is good, I took it over, but changed assert error, thank you

emob-nordic commented:
Is `config USB_CDC_ACM_DEVICE_NAME` still needed?

jfischer-no commented:
> Is `config USB_CDC_ACM_DEVICE_NAME` still needed?

no, thanks for the hint.

mbolivar-nordic commented on samples/subsys/usb/cdc_acm_composite/src/main.c:
```diff
@@ -116,13 +116,13 @@ void main(void)
 
-	dev1 = device_get_binding("CDC_ACM_1");
+	dev1 = DEVICE_DT_GET(DT_INST(1, zephyr_cdc_acm_uart));
 	if (!dev1) {
```
> I am working on a new macro `DEVICE_DT_GET_ALL(compat)`

I decided to stop at one layer beneath this macro here in this PR:

https://github.com/zephyrproject-rtos/zephyr/pull/37483

I wanted to avoid impacting device.h in case there are some compatibles where this might not be appropriate.

You should still be able use DT_FOREACH_STATUS_OKAY() here in this PR as follows:

```
#define DEVICE_AND_COMMA(node_id) DEVICE_DT_GET(node_id),
const struct *devices[] = { DT_FOREACH_STATUS_OKAY(zephyr_cdc_acm_uart, DEVICE_AND_COMMA) };
```

evgeniy-paltsev commented:
I've tested changes for ARC (both GNU and MWDT toolchains) - everything seems to be OK.

jfischer-no committed:
drivers: uart_altera_jtag_hal: use DEVICE_DT_INST_DEFINE()

The conversion to devicetree seems to be half lost
for this driver. There are already bindings and nodes for
compatible "altr,jtag-uart", update driver to use it.
Remove last mention of CONFIG_UART_CONSOLE_ON_DEV_NAME.

Resolves #37207

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
boards: native_posix: add USB device controller binding and node

Current USB device controller drivers (drivers/usb/device)
do not use DEVICE_DT_GET but the properties from devicetree,
and USB device controller node is parent for CDC ACM UART and
AUDIO children nodes.

Add USB device controller binding and node to keep the samples
building for native_posix driver.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

carlescufi commented:
@emob-nordic @mbolivar-nordic could you please take another look?

mbolivar-nordic reviewed (commented):
Excellent, looking quite good to me now, thanks. I have a few follow up questions about the changes in the latest update.

jfischer-no commented on dts/bindings/usb/zephyr,native-posix-udc.yaml:
```diff
@@ -0,0 +1,8 @@
+
+description: Native POSIX USB device controller
+
+compatible: "zephyr,native-posix-udc"
```
>I thought it was to keep the samples building, but I don't see a driver for this compatible,

Current USB device controller drivers (drivers/usb/device) do not use DEVICE_DT_GET, but the properties from devicetree. The reason is indeed to keep the samples building for native_posix even the driver (usb_dc_native_posix) is not really usable and excluded in few samples (I had a PR to reword this driver but I postponed it for the new USB device controller API).

mbolivar-nordic replied:
OK, can you amend the commit log to include that explanation please?

jfischer-no replied:
updated

jfischer-no commented on subsys/usb/class/Kconfig.cdc:
```diff
@@ -1,15 +1,27 @@
-	  USB CDC ACM device class support.
-	  Default device name is "CDC_ACM_0".
+	  USB CDC ACM class support.
+	  Default UART device name is "CDC_ACM_0".
```
No, thanks for the hint.

jfischer-no commented on samples/subsys/usb/cdc_acm_composite/src/main.c:
```diff
@@ -28,47 +28,50 @@ LOG_MODULE_REGISTER(cdc_acm_composite, LOG_LEVEL_INF);
+	struct ring_buf rb;
+};
+
+#define DEVICE_AND_COMMA(node_id) { .dev = DEVICE_DT_GET(node_id), },
```
I do not think so, it is still a device and I do not like long name/identifiers.

mbolivar-nordic replied:
It contains a device, but it is a struct serial_peer, no?

jfischer-no replied:
changed to DEFINE_SERIAL_PEER

jfischer-no commented on samples/subsys/usb/hid-cdc/src/main.c:
```diff
@@ -630,7 +630,7 @@ void main(void)
 
 	LOG_INF("Wait for DTR on CDC ACM 0");
 	while (1) {
-		uart_line_ctrl_get(cdc0_dev, UART_LINE_CTRL_DTR, &dtr);
+		uart_line_ctrl_get(cdc_dev[0], UART_LINE_CTRL_DTR, &dtr);
```
> I just want to check with you that devices 0 and 1 are interchangeable in both of these samples.

They are technically  interchangeable, but hid-cdc uses hardcoded "CDC ACM #" strings, I will check how I can fix that.

emob-nordic commented:
I think you could put some more info in release notes. Especially that cdc_acm is no longer being enabled by Kconfig option and dts overlays are used instead. You just wrote about removing `*_ON_DEV_NAME` options.

jfischer-no commented:
> Especially that cdc_acm is no longer being enabled by Kconfig option and dts overlays are used instead.

It is still enabled/disabled by Kconfig option CONFIG_USB_CDC_ACM, it is default to 'y' if a compatible node is enabled, but even then one can still disable CDC ACM class support in prj.conf file with CONFIG_USB_CDC_ACM=n.

mbolivar-nordic reviewed (approved):
All my major concerns are addressed at this point; any remaining unaddressed comments from me should be considered non-blocking.

Nice work @jfischer-no!

emob-nordic commented on samples/subsys/usb/cdc_acm_composite/src/main.c:
```diff
@@ -28,47 +28,50 @@ LOG_MODULE_REGISTER(cdc_acm_composite, LOG_LEVEL_INF);
+	DT_FOREACH_STATUS_OKAY(zephyr_cdc_acm_uart, DEFINE_SERIAL_PEER)
+};
+
+BUILD_ASSERT(ARRAY_SIZE(peers) >= 2, "Not enough CDC ACM instances");
```
Shouldn't this be asserted to be equal 2?

jfischer-no replied:
No, because there can be a node defined at the board level, e.g. for console, this example requires at least two to build.

emob-nordic commented on tests/drivers/uart/uart_mix_fifo_poll/src/main.c:
```diff
@@ -28,7 +28,7 @@
 #elif defined(CONFIG_BOARD_ATSAME54_XPRO)
```
What if `CONFIG_BOARD_ATSAME54_XPRO` is defined? The files use `DT_CHOSEN`. Will it work? If yes, then it seems those defines are not needed anymore and could be removed to avoid noise.

Same with `CONFIG_BOARD_STM32L562E_DK`

emob-nordic replied:
af2250a3c8d8585cabdb168ad5b931c0d39a7b6e this commit.

jfischer-no replied:
You are mixing things up, only uart_basic_api test uses DT_CHOSEN and DEVICE_DT_GET. For uart_mix_fifo_poll and uart_async_api, the CONFIG_UART_CONSOLE_ON_DEV_NAME is replaced by DT_LABEL(DT_CHOSEN(zephyr_console)), therefore the logic is still valid.

emob-nordic replied:
whopsie

eobalski reviewed (changes requested):
I think we still need to wait until `pull/57/head` from mcuboot is merged.

jfischer-no committed:
boards: bl654_usb: add zephyr_udc0 nodelabel

Add zephyr_udc0 (USB device controller) nodelabel to
specific USB node to allow generic USB samples to be build.

Follow up on commit e4f89478
("boards: add zephyr_udc0 nodelabel to all boards with USB support")

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
boards: bl654_usb: add chosen nodes for shell and console

Remove UART_CONSOLE_ON_DEV_NAME and UART_SHELL_ON_DEV_NAME options
and add chosen nodes for shell and console.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

nvlsianpu reviewed (changes requested):
[16:15] Puzdrowski, Andrzej
    
I see compilation error for nrf52dk_nrf52832 with this (building mcuboot):

-- Found devicetree overlay: /home/anpu/codedev/zephyrproject/bootloader/mcuboot/boot/zephyr/dts.overlay
Error: nrf52dk_nrf52832.dts.pre.tmp:559.1-13 Label or path zephyr_udc0 not found
FATAL ERROR: Syntax error parsing input tree

nvlsianpu commented:
^^ @jfischer-no Unfortunately I discovered build bug after the merge of zephyr-project/mcuboot/pull/57.

jfischer-no commented:
> ^^ @jfischer-no Unfortunately I discovered build bug after the merge of zephyr-project/mcuboot/pull/57.

https://github.com/zephyrproject-rtos/mcuboot/pull/59 should fix it.

nvlsianpu reviewed (approved):
https://github.com/zephyrproject-rtos/mcuboot/pull/59 fixes the build failure I mentioned.
